core/src/main/java/org/ldaptive/url/Url.java

```diff
@@ -0,0 +1,443 @@
+/* See LICENSE for licensing and NOTICE for copyright. */
+package org.ldaptive.url;
+
+import java.util.Arrays;
+import org.ldaptive.LdapUtils;
+import org.ldaptive.SearchScope;
+import org.ldaptive.dn.Dn;
+import org.ldaptive.filter.Filter;
+import org.ldaptive.filter.PresenceFilter;
+
+/**
+ * Class to represent an LDAP URL. See RFC 4516. Expects URLs of the form
+ * scheme://hostname:port/baseDn?attrs?scope?filter. This implementation does not support URL extensions.
+ *
+ * @author  Middleware Services
+ */
+public final class Url
+{
+
+  /** Default LDAP port, value is {@value}. */
+  static final int DEFAULT_LDAP_PORT = 389;
+
+  /** Default LDAPS port, value is {@value}. */
+  static final int DEFAULT_LDAPS_PORT = 636;
+
+  /** Default LDAPI port, value is {@value}. */
+  static final int DEFAULT_LDAPI_PORT = 0;
+
+  /** Default base DN, value is {@value}. */
+  static final String DEFAULT_BASE_DN = "";
+
+  /** Default parsed base DN. */
+  static final Dn DEFAULT_PARSED_BASE_DN = new Dn();
+
+  /** Default attribute filter value is '(objectClass=*)'. */
+  static final String DEFAULT_FILTER = "(objectClass=*)";
+
+  /** Default parsed attribute filter. */
+  static final Filter DEFAULT_PARSED_FILTER = new PresenceFilter("objectClass");
+
+  /** Default scope, value is {@link SearchScope#OBJECT}. */
+  static final SearchScope DEFAULT_SCOPE = SearchScope.OBJECT;
+
+  /** Default return attributes, value is all user attributes. */
+  static final String[] DEFAULT_ATTRIBUTES = new String[0];
+
+  /** hash code seed. */
+  private static final int HASH_CODE_SEED = 11003;
+
+  /** Scheme of the ldap url. */
+  private final String scheme;
+
+  /** Hostname of the ldap url. */
+  private final String hostname;
+
+  /** Port of the ldap url. */
+  private final int port;
+
+  /** Base DN of the ldap url. */
+  private final String baseDn;
+
+  /** Parsed base DN of the ldap url. */
+  private final Dn parsedBaseDn;
+
+  /** Attributes of the ldap url. */
+  private final String[] attributes;
+
+  /** Search scope of the ldap url. */
+  private final SearchScope scope;
+
+  /** Search filter of the ldap url. */
+  private final String filter;
+
+  /** Parsed filter of the ldap url. */
+  private final Filter parsedFilter;
```

Comment:
Same concern as DN. It's ok to represent both forms internally as an optimization/convenience/implementation detail, but both forms should not be exposed to construct the object.

Reply:
Resolving this as part of the `parsedDn` conversation.